File: java-srcs/src/main/java/com/samoylov/leetcode/easy/RangeSumOfBST.java
package com.samoylov.leetcode.easy;

/**
 * <a href="https://leetcode.com/problems/range-sum-of-bst/description/">task link</a>
 */
public class RangeSumOfBST {
    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     * int val;
     * TreeNode left;
     * TreeNode right;
     * TreeNode() {}
     * TreeNode(int val) { this.val = val; }
     * TreeNode(int val, TreeNode left, TreeNode right) {
     * this.val = val;
     * this.left = left;
     * this.right = right;
     * }
     * }
     */
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int rangeSumBST(TreeNode root, int low, int high) {
        if (root == null) {
            return 0;
        }
        return rangeSum(root, low, high);
    }

    public int rangeSum(TreeNode node, int low, int high) {
        var result = 0;
        if (node.val <= high && node.val >= low) {
            result += node.val;
        }
        if (node.left != null) {
            result += rangeSum(node.left, low, high);
        }
        if (node.right != null) {
            result += rangeSum(node.right, low, high);
        }
        return result;
    }
}

File: go-srcs/leetcode/easy/range_sum_of_bst.go
package easy

// task link - https://leetcode.com/problems/range-sum-of-bst/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rangeSumBST(root *TreeNode, low int, high int) int {
	var sum int

	countRangeSum(root, &sum, low, high)

	return sum
}

func countRangeSum(cur *TreeNode, sum *int, low int, high int) {
	if cur == nil {
		return
	}
	if cur.Val >= low && cur.Val <= high {
		*sum += cur.Val
	}
	countRangeSum(cur.Left, sum, low, high)
	countRangeSum(cur.Right, sum, low, high)
}
